tests/test_nonbonded_interaction_group.py

```diff
@@ -0,0 +1,129 @@
+import jax
+
+jax.config.update("jax_enable_x64", True)
+
+import numpy as np
+import pytest
+from common import GradientTest, prepare_reference_nonbonded
+from simtk.openmm import app
+
+from timemachine.fe.utils import to_md_units
+from timemachine.ff.handlers import openmm_deserializer
+from timemachine.lib import potentials
+from timemachine.lib.potentials import NonbondedInteractionGroup
+
+
+@pytest.fixture(autouse=True)
+def set_random_seed():
+    np.random.seed(2021)
+    yield
+
+
+@pytest.fixture
+def test_system():
+    pdb_path = "tests/data/5dfr_solv_equil.pdb"
+    host_pdb = app.PDBFile(pdb_path)
+    ff = app.ForceField("amber99sbildn.xml", "tip3p.xml")
+    return (
+        ff.createSystem(host_pdb.topology, nonbondedMethod=app.NoCutoff, constraints=None, rigidWater=False),
+        host_pdb.positions,
+        host_pdb.topology.getPeriodicBoxVectors(),
+    )
+
+
+@pytest.fixture
+def ref_nonbonded_potential(test_system):
+    host_system, _, _ = test_system
+    host_fns, _ = openmm_deserializer.deserialize_system(host_system, cutoff=1.0)
+
+    nonbonded_fn = None
+    for f in host_fns:
+        if isinstance(f, potentials.Nonbonded):
+            nonbonded_fn = f
+
+    return nonbonded_fn
+
+
+@pytest.fixture
+def test_conf(test_system):
+    _, host_coords, _ = test_system
+    return np.array([[to_md_units(x), to_md_units(y), to_md_units(z)] for x, y, z in host_coords])
+
+
+@pytest.fixture
+def test_box(test_system):
+    _, _, box = test_system
+    return np.asarray(box / box.unit)
+
+
+@pytest.mark.parametrize("beta", [2.0])
+@pytest.mark.parametrize("cutoff", [1.1])
+@pytest.mark.parametrize("precision,rtol,atol", [(np.float64, 1e-8, 1e-8), (np.float32, 1e-4, 5e-4)])
+@pytest.mark.parametrize("num_row_atoms", [1, 15])
+@pytest.mark.parametrize("num_atoms", [33, 231, 1050])
+def test_nonbonded_interaction_group_correctness(
```

Comment:
A couple other possible self-consistency tests:

* Modify coords in a way that influences U_AB but not U_A or U_B, something like:
```python
# ... 
conf_prime = np.array(conf)
conf_prime[ligand_idxs] += 0.01 * np.random.randn(len(ligand_idxs), 3)  # assuming U_A is shift-invariant

ref_delta_U = U_ref(conf_prime) - U_ref(conf)
test_delta_U = U_ixn(conf_prime) - U_ixn(conf)
# ...
# assert ref_delta_U close to test_delta_U
```

* Modify lam in a way that influences U_AB but not U_A or U_B.
(For example, construct with lambda planes / offsets all one value for atoms in A, and a different value for all atoms in B.)

Reply:
I find modification of lambda values to test this quite confusing. It seems like we would want a generic jax reference implementation of this feature to test it?

Reply:
>I find modification of lambda values to test this quite confusing. 

In absolute free energy "decoupling" stages, do we not already define the lambda planes / offsets in such a way that `lam` only controls ligand-environment interactions (but does not influence ligand-ligand or environment-environment interactions, `U(x, lam) = U_A(x) + U_B(x) + U_AB(x, lam)`)?

If so, this would be another possible route to a self-consistency test for a new way to compute `U_AB`.

>It seems like we would want a generic jax reference implementation of this feature to test it?

Also a good idea to compare to jax reference implementations.

A helpful starting point may be this timing script using `timemachine.potentials.jax_utils` https://github.com/proteneer/timemachine/pull/545#issuecomment-991190069

Reply:
I added an additional consistency check comparing with [`nonbonded_v3_on_specific_pairs`](https://github.com/proteneer/timemachine/blob/5d8f3099aafa1cf7def0585525aaf4de73c23343/timemachine/potentials/nonbonded.py#L222) in eb07db3. This is hopefully more straightforward to follow.

Interestingly, I am getting a failure for this new check for some sizes, and seeing that only a small number of du_dps are significantly off. I will follow up on this.

Reply:
> A couple other possible self-consistency tests:

Nice, thanks for the suggestions! I'll look into adding these next.

Reply:
> Interestingly, I am getting a failure for this new check for some sizes

I think I've tracked down the cause of this -- the nonbonded kernel [returns du/dp=0 for Lennard-Jones parameters at zero](https://github.com/proteneer/timemachine/blob/39907d6cc2bfe9ba36d4fffdecdeb93889507103/timemachine/cpp/src/kernels/k_nonbonded.cuh#L420), but the reference jax implementation returns nonzero values. Worked around this in 39907d6.

Reply:
>I think I've tracked down the cause of this -- the nonbonded kernel [returns du/dp=0 for Lennard-Jones parameters at zero](https://github.com/proteneer/timemachine/blob/39907d6cc2bfe9ba36d4fffdecdeb93889507103/timemachine/cpp/src/kernels/k_nonbonded.cuh#L420), but the reference jax implementation returns nonzero values. Worked around this in https://github.com/proteneer/timemachine/commit/39907d6cc2bfe9ba36d4fffdecdeb93889507103.

Ahh -- I had overlooked this...

Tangents:
1. The reference jax implementation should probably be modified to stop gradient when `eps_ij == 0` https://github.com/proteneer/timemachine/blob/39907d6cc2bfe9ba36d4fffdecdeb93889507103/timemachine/potentials/nonbonded.py#L177  --> maybe by adding `eps_ij != 0` condition to `keep_mask` https://github.com/proteneer/timemachine/blob/39907d6cc2bfe9ba36d4fffdecdeb93889507103/timemachine/potentials/nonbonded.py#L208 
2. The special case of `eps_i == eps_j == 0` should be exercised in the default nonbonded correctness test, if it's not already

Reply:
Ahh, actually this line can just be deleted https://github.com/proteneer/timemachine/blob/39907d6cc2bfe9ba36d4fffdecdeb93889507103/timemachine/potentials/nonbonded.py#L208 (it overwrites a `keep_mask` that already reflects the `eps_ij != 0` condition)

Reply:
> Ahh, actually this line can just be deleted

This should be addressed by 860dc9d (we don't use `nonbonded_v3` for the test in question, but a similar change was needed in `nonbonded_v3_specific_pairs`)

Reply:
Ahh, sorry for these bunch of red herrings -- the value of `keep_mask` assigned at L182 applies to vdw, the value of `keep_mask` assigned at L208 applies to coulomb... (And as @mcwitt pointed out offline: this whole function isn't involved in the relevant test!)

Reply:
(Marking resolved)